src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -2612,110 +2741,290 @@ AdapterHandlerEntry* AdapterHandlerLibrary::get_adapter(const methodHandle& meth
       }
     }
+    return true;
+  }
+};
+
+void AdapterHandlerLibrary::dump_aot_adapter_table() {
+  CompactHashtableStats stats;
+  CompactHashtableWriter writer(_adapter_handler_table->number_of_entries(), &stats);
+  CopyAdapterTableToArchive copy(&writer);
+  _adapter_handler_table->iterate(&copy);
+  writer.dump(&_aot_adapter_handler_table, "archived adapter table");
+}
+
+void AdapterHandlerLibrary::serialize_shared_table_header(SerializeClosure* soc) {
+  _aot_adapter_handler_table.serialize_header(soc);
+}
+
+AdapterBlob* AdapterHandlerLibrary::link_aot_adapter_handler(AdapterHandlerEntry* handler) {
+#ifdef ASSERT
+  if (TestAOTAdapterLinkFailure) {
+    return nullptr;
   }
 #endif
+  AdapterBlob* blob = lookup_aot_cache(handler);
+#ifndef PRODUCT
+  // debugging support
+  if ((blob != nullptr) && (PrintAdapterHandlers || PrintStubCode)) {
+    print_adapter_handler_info(handler, blob);
+  }
+#endif
+  return blob;
+}
 
-  // Add the entry only if the entry contains all required checks (see sharedRuntime_xxx.cpp)
-  // The checks are inserted only if -XX:+VerifyAdapterCalls is specified.
-  if (contains_all_checks || !VerifyAdapterCalls) {
-    assert_lock_strong(AdapterHandlerLibrary_lock);
-    _adapter_handler_table->put(fingerprint, entry);
+// This method is used during production run to link archived adapters (stored in AOT Cache)
+// to their code in AOT Code Cache
+void AdapterHandlerEntry::link() {
+  AdapterBlob* adapter_blob = nullptr;
+  ResourceMark rm;
+  assert(_fingerprint != nullptr, "_fingerprint must not be null");
+  bool generate_code = false;
+  // Generate code only if AOTCodeCache is not available, or
+  // caching adapters is disabled, or we fail to link
+  // the AdapterHandlerEntry to its code in the AOTCodeCache
+  if (AOTCodeCache::is_using_adapters()) {
+    adapter_blob = AdapterHandlerLibrary::link_aot_adapter_handler(this);
+    if (adapter_blob == nullptr) {
+      log_warning(cds)("Failed to link AdapterHandlerEntry (fp=%s) to its code in the AOT code cache", _fingerprint->as_basic_args_string());
+      generate_code = true;
+    }
+  } else {
+    generate_code = true;
+  }
+  if (generate_code) {
+    int nargs;
+    BasicType* bt = _fingerprint->as_basic_type(nargs);
+    if (!AdapterHandlerLibrary::generate_adapter_code(adapter_blob, this, nargs, bt, /* is_transient */ false)) {
+      // Don't throw exceptions during VM initialization because java.lang.* classes
+      // might not have been initialized, causing problems when constructing the
+      // Java exception object.
+      vm_exit_during_initialization("Out of space in CodeCache for adapters");
+    }
   }
-  return entry;
+  // Outside of the lock
+  if (adapter_blob != nullptr) {
+    post_adapter_creation(adapter_blob, this);
+  }
+  assert(_linked, "AdapterHandlerEntry must now be linked");
+}
+
+void AdapterHandlerLibrary::link_aot_adapters() {
```

Comment:
I am wondering if we need to acquire `AdapterHandlerLibrary_lock`, as done by other methods called during this stage (eg `lookup_simple_adapters` and `AdapterHandlerLibrary::initialize`). 
Linking of adapters happens very early in bootstrap, where we have only one thread. But this lock was acquired in earlier code version of the code during bootstrap as well. So to keep things consistent, I feel we should hold the lock here.

Reply:
Please, give me patch where we should place locks.

Reply:
@vnkozlov here is the patch for adding the lock: https://github.com/ashu-mehra/jdk/commit/a1c463ddc85e7c179b32ce1c3f209f623cb6db71
It also adds back PerfTraceTimer that I missed in earlier refactoring.